### PR TITLE
Add method to note transaction unconfirmed/reorged-out

### PR DESCRIPTION
**Pull request by TheBlueMatt**
This is based on #838, and adds two new methods `transactions_unconfirmed()` and `get_relevant_txids()` to `ChannelManager`, allowing an electrum client to explicitly notify us that the funding transaction has been reorganized out of the chain. Needs more docs and a more careful analysis of whether we can make this API work for ChannelMonitor, but for now we may ship a Java bindings binary with this.

### CODE REVIEW

File: lightning/src/ln/reorg_tests.rs [hunk 1]
```diff
@@ -196,16 +196,28 @@ fn do_test_unconf_chan(reload_node: bool, reorg_after_reload: bool, connect_styl
...
 	assert_eq!(channel_state.short_to_id.len(), 1);
 	mem::drop(channel_state);
 
 	if !reorg_after_reload {
-		disconnect_all_blocks(&nodes[0]);
-		check_closed_broadcast!(nodes[0], true);
+		if use_funding_unconfirmed {
+			let relevant_txids = nodes[0].node.get_relevant_txids();
+			assert_eq!(relevant_txids.len(), 1);
+			assert_eq!(&relevant_txids[..], &[chan.3.txid()]);
```
Comment (jkczyz) [inline lightning/src/ln/reorg_tests.rs:210]: Compare `relevant_txids[0]` directly?

File: lightning/src/ln/reorg_tests.rs [hunk 2]
```diff
@@ -196,16 +196,28 @@ fn do_test_unconf_chan(reload_node: bool, reorg_after_reload: bool, connect_styl
...
+			nodes[0].node.transaction_unconfirmed(&relevant_txids[0]);
+		} else {
+			disconnect_all_blocks(&nodes[0]);
+		}
+		if connect_style == ConnectStyle::FullBlockViaListen && !use_funding_unconfirmed {
+			handle_announce_close_broadcast_events(&nodes, 0, 1, true, "Funding transaction was un-confirmed. Locked at 6 confs, now have 2 confs.");
+		} else {
+			handle_announce_close_broadcast_events(&nodes, 0, 1, true, "Funding transaction was un-confirmed. Locked at 6 confs, now have 0 confs.");
+		}
+		check_added_monitors!(nodes[1], 1);
```
Comment (jkczyz) [inline lightning/src/ln/reorg_tests.rs:220]: This test keeps growing... Would it make sense to separate the various behaviors into their own tests at this point? It isn't clear to me that the cartesian product needs to be tested.

**Review by jkczyz — COMMENTED** [review]
Concept ACK. Will investigate the `ChannelMonitor` side.

File: lightning/src/ln/channelmanager.rs [hunk 3]
```diff
@@ -3477,6 +3479,30 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
...
+		res
+	}
+
+	/// Marks a transaction as having been reorganized out of the blockchain XXX: Docs
+	pub fn transaction_unconfirmed(&self, txid: &Txid) {
+		let _persistence_guard = PersistenceNotifierGuard::new(&self.total_consistency_lock, &self.persistence_notifier);
+		self.do_chain_event(None, |channel| {
+			if let Some(funding_txo) = channel.get_funding_txo() {
+				if funding_txo.txid == *txid {
+					channel.funding_transaction_unconfirmed().map(|_| (None, Vec::new()))
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:3505]: Hmm, we could probably macro-ize it, but `do_chain_event` handles the returned values, too, so I think they need to be concrete somehow in the body of the function?

File: lightning/src/ln/channelmanager.rs [hunk 4]
```diff
@@ -3477,6 +3479,30 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 		}
 	}
 
+	/// Gets the set of txids which should be monitored for their confirmation state. XXX: Docs
+	pub fn get_relevant_txids(&self) -> Vec<Txid> {
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:3488]: In the `ChannelManager` context, yes, but it wouldn't work for `ChannelMonitor` - `ChannelManager` only ever cares about one, known-in-advance-txid transaction, whereas `ChannelMonitor` doesn't know about txids that it cares about until they are confirmed (or, it shouldn't expose them, the list is potentially huge).

File: lightning/src/ln/channel.rs [hunk 5]
```diff
@@ -3675,6 +3675,24 @@ impl<Signer: Sign> Channel<Signer> {
 		Ok((None, timed_out_htlcs))
 	}
 
+	pub fn funding_transaction_unconfirmed(&mut self) -> Result<(), msgs::ErrorMessage> {
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs:3678]: `channel.rs` itself is already only `pub(crate)`, so `pub` is used rather extensively. I can switch them if you prefer, it doesn't matter either way.

File: lightning/src/ln/reorg_tests.rs (hunk 2 above)
Comment (TheBlueMatt) [reply lightning/src/ln/reorg_tests.rs:220]: Hmm, we could clone the test and run the full set of connection styles against something that drops the reload code, but we'd still end up with basically the same thing duplicated. I definitely think we need to test the set of full connection/disconnection styles, since I found unexpected bugs in this test when doing so across the full test suite :).

**Comment by TheBlueMatt** [comment]
Rebased now that #838 has been merged.

File: lightning/src/ln/channel.rs [hunk 6]
```diff
@@ -3675,6 +3675,24 @@ impl<Signer: Sign> Channel<Signer> {
...
+		if self.funding_tx_confirmation_height != 0 {
+			let effective_new_height = self.funding_tx_confirmation_height - 1;
```
Comment (ariard) [inline lightning/src/ln/channel.rs:3680]: IIUC, setting "effective_new_height" to "funding_tx_confirmation_height" - 1 will force this latter value to be reset to 0 in `check_get_funding_locked` ? I don't think "effective_new_height" is a compelling name if so maybe "unconf_height" and maybe explain with a code comment that it will latch "funding_tx_confirmation_height" to 0, the effect is 2 function calls under.

File: lightning/src/ln/channelmanager.rs [hunk 7]
```diff
@@ -850,6 +850,11 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 		}
 	}
 
+	/// Gets the current configuration applied to all new channels,  as
+	pub fn get_current_default_configuration(&self) -> &UserConfig {
```
Comment (ariard) [inline lightning/src/ln/channelmanager.rs:854]: Need to correct comment, "as" what ?

File: lightning/src/ln/channel.rs (hunk 5 above)
Comment (ariard) [inline lightning/src/ln/channel.rs:3678]: doc ?

**Review by ariard — COMMENTED** [review]
SGTM modulo small comments

**Comment by TheBlueMatt** [comment]
Addressed comments and added requisite documentation, this should be ready to go now.

File: lightning/src/ln/channelmanager.rs [hunk 8]
```diff
@@ -3522,6 +3529,54 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
...
+		let mut res = Vec::with_capacity(channel_state.short_to_id.len());
+		for chan in channel_state.by_id.values() {
+			if let Some(funding_txo) = chan.get_funding_txo() {
+				res.push(funding_txo.txid);
+			}
+		}
+		res
+	}
+
+	/// Marks a transaction as having been reorganized out of the blockchain.
```
Comment (valentinewallace) [inline lightning/src/ln/channelmanager.rs:3553]: I'm confused because this indicates the transaction was def reorged out, but the docs for `channel.transaction_unconfirmed` say "...may also indicate a harmless reorganization of a block or two."? Seems to be a contradiction?

File: lightning/src/ln/channel.rs [hunk 9]
```diff
@@ -3678,6 +3678,28 @@ impl<Signer: Sign> Channel<Signer> {
 		Ok((None, timed_out_htlcs))
 	}
 
+	/// Indicates the funding transaction is no longer confirmed in the main chain. This may
+	/// force-close the channel, but may also indicate a harmless reorganization of a block or two.
+	pub fn funding_transaction_unconfirmed(&mut self) -> Result<(), msgs::ErrorMessage> {
+		if self.funding_tx_confirmation_height != 0 {
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs:3684]: Hmm, we do reset it back to zero in `update_best_block` in the `funding_tx_confirmation_height == 0` check.

File: lightning/src/ln/channelmanager.rs [hunk 10]
```diff
@@ -3522,6 +3529,54 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
...
+	/// Gets the set of txids which should be monitored for their confirmation state.
+	///
+	/// If you're providing information about reorganizations via [`transaction_unconfirmed`], this
+	/// is the set of transactions which you may been to call [`transaction_unconfirmed`] for.
+	///
+	/// Note that this is NOT the set of transactions which must be included in calls to
+	/// [`transactions_confirmed`] if they are confirmed, but a small subset of it.
+	///
+	/// [`transactions_confirmed`]: Self::transactions_confirmed
+	/// [`transaction_unconfirmed`]: Self::transaction_unconfirmed
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:3547]: Tried to address this, let me know if that's clearer.

File: lightning/src/ln/channelmanager.rs (hunk 8 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:3559]: Its referring to the case where we didn't send a funding_locked yet. I rephrased it.

File: lightning/src/ln/channelmanager.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (jkczyz) [reply lightning/src/ln/channelmanager.rs:3547]: FWIW, I'll try to unify and clean-up the docs for this and those in #858 when making a trait for these to implement.

File: lightning/src/ln/channelmanager.rs [hunk 11]
```diff
@@ -3527,7 +3527,13 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
...
 	///
 	/// If you're providing information about reorganizations via [`transaction_unconfirmed`], this
-	/// is the set of transactions which you may been to call [`transaction_unconfirmed`] for.
+	/// is the set of transactions which you may need to call [`transaction_unconfirmed`] for.
+	///
+	/// This may be useful to poll to determine the set of transactions which must be registered
+	/// with an Electrum server or for which an electrum server needs to be polled to determine
+	/// transaction confirmation state.
+	///
+	/// This may update after any [`transactions_confirmed`] or [`block_connected`] call.
```
Comment (jkczyz) [inline lightning/src/ln/channelmanager.rs:3541]: What's the use case for calling this with `block_connected`?

File: lightning/src/ln/channel.rs [hunk 12]
```diff
@@ -3679,7 +3679,8 @@ impl<Signer: Sign> Channel<Signer> {
 	}
 
 	/// Indicates the funding transaction is no longer confirmed in the main chain. This may
-	/// force-close the channel, but may also indicate a harmless reorganization of a block or two.
+	/// force-close the channel, but may also indicate a harmless reorganization of a block or two
+	/// before the channel has reached funding_locked and we can just wait for more blocks.
```
Comment (jkczyz) [inline lightning/src/ln/channel.rs]: Second sentence is a bit wordy / run-on. But isn't it really the caller that will close the channel upon receiving an error back?

File: lightning/src/ln/channel.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (valentinewallace) [reply lightning/src/ln/channel.rs:3685]: Hmm, that doesn't reset `self.funding_tx_confirmation_height` I think? https://i.imgur.com/kgXgXLV.png

File: lightning/src/ln/channel.rs [hunk 13]
```diff
@@ -3678,6 +3678,29 @@ impl<Signer: Sign> Channel<Signer> {
...
+				Ok((funding_locked, timed_out_htlcs)) => {
+					assert!(funding_locked.is_none(), "We can't generate a funding with 0 confirmations?");
+					assert!(timed_out_htlcs.is_empty(), "We can't have accepted HTLCs with a timeout before our funding confirmation?");
+					Ok(())
+				},
+				Err(e) => Err(e)
+			}
+		} else {
+			// We never learned about the funding confirmation anyway, just ignore
+			Ok(())
```
Comment (valentinewallace) [inline lightning/src/ln/channel.rs:3700]: Is this theoretically supposed to be unreachable, but it seems harmless so we can just `Ok(())` in practice?

File: lightning/src/ln/channel.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (valentinewallace) [reply lightning/src/ln/channel.rs:3685]: Also, should we be resetting `funding_tx_confirmed_in` and `short_channel_id` as well? (And maybe unit test this if it's easy?)

File: lightning/src/ln/channel.rs [hunk 14]
```diff
@@ -3678,6 +3678,29 @@ impl<Signer: Sign> Channel<Signer> {
...
 
+	/// Indicates the funding transaction is no longer confirmed in the main chain. This may
+	/// force-close the channel, but may also indicate a harmless reorganization of a block or two
+	/// before the channel has reached funding_locked and we can just wait for more blocks.
+	pub fn funding_transaction_unconfirmed(&mut self) -> Result<(), msgs::ErrorMessage> {
+		if self.funding_tx_confirmation_height != 0 {
+			// We handle the funding disconnection by calling update_best_block with a height one
+			// below where our funding was connected, implying a reorg back to conf_height - 1.
+			let reorg_height = self.funding_tx_confirmation_height - 1;
+			let best_time = self.update_time_counter;
```
Comment (valentinewallace) [inline lightning/src/ln/channel.rs:3689]: Maybe a comment for why this stays the same? Seems like since it's just for a timestamp, it'd still make sense to be incremented..

File: lightning/src/ln/channelmanager.rs [hunk 15]
```diff
@@ -3365,7 +3370,7 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
         L::Target: Logger,
 {
 	fn do_chain_event<FN: Fn(&mut Channel<Signer>) -> Result<(Option<msgs::FundingLocked>, Vec<(HTLCSource, PaymentHash)>), msgs::ErrorMessage>>
```
Comment (valentinewallace) [inline lightning/src/ln/channelmanager.rs:3372]: I think a high-level comment could be useful for this function

File: lightning/src/ln/channelmanager.rs [hunk 16]
```diff
@@ -3522,6 +3529,60 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 		}
 	}
 
+	/// Gets the set of txids which should be monitored for their confirmation state.
+	///
+	/// If you're providing information about reorganizations via [`transaction_unconfirmed`], this
+	/// is the set of transactions which you may need to call [`transaction_unconfirmed`] for.
+	///
+	/// This may be useful to poll to determine the set of transactions which must be registered
+	/// with an Electrum server or for which an electrum server needs to be polled to determine
```
Comment (valentinewallace) [inline lightning/src/ln/channelmanager.rs:3538]: nit: s/electrum/Electrum

File: lightning/src/ln/channelmanager.rs [hunk 17]
```diff
@@ -3522,6 +3529,60 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
...
+	/// transaction confirmation state.
+	///
+	/// This may update after any [`transactions_confirmed`] or [`block_connected`] call.
+	///
+	/// Note that this is NOT the set of transactions which must be included in calls to
+	/// [`transactions_confirmed`] if they are confirmed, but a small subset of it.
```
Comment (valentinewallace) [inline lightning/src/ln/channelmanager.rs:3544]: I assume the rest will be registered in one of the `register_tx` calls? Could be worth adding a note, since this line may come across naively as a bit scary but unactionable...

Nw though if Jeff is addressing the docs more thoroughly in #858.

File: lightning/src/ln/channelmanager.rs [hunk 18]
```diff
@@ -3522,6 +3529,60 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
...
+	/// If a transaction is included in [`get_relevant_txids`], and is no longer in the main branch
+	/// of the blockchain, this function should be called to indicate that the transaction should
+	/// be considered reorganized out.
+	///
+	/// Once this is called, the given transaction will no longer appear on [`get_relevant_txids`],
+	/// though this may be called repeatedly for a given transaction without issue.
+	///
+	/// Note that if the transaction is confirmed on the main chain in a different block (indicated
+	/// via a call to [`transactions_confirmed`]), it may re-appear in [`get_relevant_txids`], thus
+	/// be very weary of race-conditions wherein the final state of a transaction indicated via
```
Comment (valentinewallace) [inline lightning/src/ln/channelmanager.rs:3570]: nit: s/weary/wary

File: lightning/src/ln/channelmanager.rs [hunk 19]
```diff
@@ -3522,6 +3529,60 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
...
+	/// these APIs is not the same as its state on the blockchain.
```
Comment (valentinewallace) [inline lightning/src/ln/channelmanager.rs:3571]: Found this race condition description a bit confusing -- I assume "these APIs" == `transaction_confirmed`/`get_relevant_txids`? 

"the final state of a transaction indicated by these APIs is not the same as its state on the blockchain" == just because a tx is included or not included in `get_relevant_txids`, doesn't tell you anything about this state on the blockchain? If that's correct, then sure, but I guess it's not obvious to me how that'd lead to a race condition.

File: lightning/src/ln/channel.rs (hunk 13 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs:3700]: Yea, I think it should be really rare, but you could see some async process getting delayed and ending up calling `transaction_unconfirmed` twice in a row which could hit this.

File: lightning/src/ln/channel.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs:3685]: Currently we'll always close the channel in this case due to the `if funding_tx_confirmations < self.minimum_depth as i64 / 2 {` check. I'll add a test and stop doing that in a followup.

File: lightning/src/ln/channelmanager.rs (hunk 18 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:3570]: TIL.

File: lightning/src/ln/channelmanager.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:3547]: Yep, sounds good. good to have an example here to pull from/merge from, too.

File: lightning/src/ln/channelmanager.rs (hunk 17 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:3544]: I believe the plan is still for @jkczyz to do a followup PR after this + #858 that will add a new top-level trait similar to `chain::Listen` and merge all the docs/functions then.

**Comment by TheBlueMatt** [comment]
Addressed a number of doc comments, but for the top-level docs on `transaction_unconfirmed`, lets plan on cleaning those up in a followup? Since they'll move to a trait/merge with #858 anyway.

File: lightning/src/ln/channelmanager.rs (hunk 7 above)
Comment (jkczyz) [reply lightning/src/ln/channelmanager.rs:855]: Yeah, still needs to be fixed.

**Review by jkczyz — APPROVED** [review]

**Review by valentinewallace — APPROVED** [review]

**Comment by TheBlueMatt** [comment]
Squashed with one doc-bug issue CI pointed out:

```
$ git diff-tree -U1 b6a8560a 6982434c
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index 81cd95ef..f9e00bc5 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -3550,2 +3550,3 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 	/// [`transaction_unconfirmed`]: Self::transaction_unconfirmed
+	/// [`block_connected`]: chain::Listen::block_connected
 	pub fn get_relevant_txids(&self) -> Vec<Txid> {
```
